sc_demographics.R

```diff
@@ -0,0 +1,280 @@
+  "L15 0PR",
+  "M16 0GS",
+  "ML3 0GS",
+  "ML7 6AQ",
+  "NK1 1AA",
+  "NK1 1AB",
+  "PA438JP",
+  "PA494JS",
+  "PR2 5AL",
+  "TD8 8JD"
+)
+
+sc_demog <-
+  sc_demog %>%
+  mutate(
+    # remove dummy postcodes "NK1 0AA" and "NK1 1AB"
+    submitted_postcode = na_if(
+      submitted_postcode,
+      submitted_postcode == "NK1 0AA" | submitted_postcode == "NF1 1AB"
+    ),
+    chi_postcode = na_if(
+      chi_postcode,
+      chi_postcode == "NK1 0AA" | chi_postcode == "NF1 1AB"
+    )
+  ) %>%
+  mutate(
+    # comparing with regex UK postcode
+    submitted_postcode = na_if(submitted_postcode, !str_detect(
+      submitted_postcode,
+      "([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))\\s?[0-9][A-Za-z]{2})"
+    )),
+    chi_postcode = na_if(chi_postcode, !str_detect(
+      chi_postcode,
+      "([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))\\s?[0-9][A-Za-z]{2})"
+    ))
+  ) %>%
+  mutate(
+    # na if postcode is in the invalid postcode vector
+    submitted_postcode = na_if(submitted_postcode, submitted_postcode %in% invalid_postcodes),
+    chi_postcode = na_if(chi_postcode, chi_postcode %in% chi_postcode)
+  ) %>%
+  # sort by submitted_postcode
+  arrange(submitted_postcode)
+
+
+
+## postcode type ##
+postcode_lookup <- readr::read_rds(file = "/conf/linkage/output/lookups/Unicode/Geography/Scottish Postcode Directory/Scottish_Postcode_Directory_2021_2.rds")
+
+sc_demog <-
+  sc_demog %>%
+  select(
+    latest_record_flag, extract_date, sending_location, social_care_id, upi, gender, dob, postcode,
+    submitted_postcode, chi_postcode
+  ) %>%
+  # check if submitted_postcode matches with postcode lookup
+  mutate(valid_pc = if_else(submitted_postcode %in% postcode_lookup$pc7, 1, 0))
+
+# use submitted_postcode if valid, otherwise use chi_postcode
+sc_demog <-
+  sc_demog %>%
+  mutate(postcode = case_when(
+    (!is.na(submitted_postcode) & valid_pc == 1) ~ submitted_postcode,
+    (is.na(submitted_postcode) & valid_pc == 0) ~ chi_postcode
+  )) %>%
+  mutate(postcode_type = case_when(
+    (!is.na(submitted_postcode) & valid_pc == 1) ~ "submitted",
+    (is.na(submitted_postcode) & valid_pc == 0) ~ "chi",
+    (is.na(submitted_postcode) & is.na(chi_postcode)) ~ "missing"
+  ))
+
+
+# Check where the postcodes are coming from
+sc_demog %>%
+  count(postcode_type)
+
+
+## outfile ##
+outfile <-
+  sc_demog %>%
+  # sort so latest submissions are last
+  arrange(sending_location, social_care_id, latest_record_flag, extract_date) %>%
+  ## Aggregate to create one row per sending_location / ID ##
+  distinct(social_care_id, .keep_all = TRUE) %>%
+  mutate(
+    sending_location = coalesce(sending_location),
+    chi = coalesce(upi),
+    gender = coalesce(gender),
+    dob = coalesce(dob),
+    postcode = coalesce(postcode)
+  ) %>%
```

Comment:
I don't think this will do anything, what was the intention here?

Reply:
@Moohan I used the similar idea from the Deaths Extract to create one row per sending location / ID - is this not similar? 

The spss code uses last() so I thought coalesce was similar?